### PR TITLE
Refactored use of formatter factory

### PR DESCRIPTION
**Pull request by ssbarnea**
- reduce branching complexity by moving the formatter_factory into a separate function
- removed linting exception for main complexity
- add typing information to formatter classes

### CODE REVIEW

File: lib/ansiblelint/__main__.py [hunk 1]
```diff
@@ -56,17 +57,16 @@ def initialize_logger(level: int = 0) -> None:
 
 def choose_formatter_factory(
     options_list: "Namespace"
-) -> Union[Type[formatters.Formatter], Type[formatters.QuietFormatter],
-           Type[formatters.ParseableFormatter],
-           Type[formatters.ParseableSeverityFormatter]]:
+) -> Type[formatters.BaseFormatter]:
```
Comment (webknjaz) [inline lib/ansiblelint/__main__.py:60]: I believe that the TypeVar approach would be more appropriate here.

File: lib/ansiblelint/__main__.py [hunk 2]
```diff
@@ -56,6 +59,20 @@ def initialize_logger(level: int = 0) -> None:
...
+    """Select an output formatter based on the incoming command line arguments."""
+    r: Type[formatters.BaseFormatter] = formatters.Formatter
+    if options_list.quiet:
+        r = formatters.QuietFormatter
+    elif options_list.parseable:
+        r = formatters.ParseableFormatter
+    elif options_list.parseable_severity:
+        r = formatters.ParseableSeverityFormatter
+    return r
```
Comment (webknjaz) [inline lib/ansiblelint/__main__.py:73]: Extra var is unnecessary but instead using a TypeVar is needed as requested @ https://github.com/ansible/ansible-lint/pull/931/files#r463479262.

File: lib/ansiblelint/formatters/__init__.py [hunk 3]
```diff
@@ -1,15 +1,17 @@
...
 from ansiblelint.color import Color, colorize
 
 if TYPE_CHECKING:
     from ansiblelint.errors import MatchError
 
+T = TypeVar('T', bound='BaseFormatter')
 
-class BaseFormatter:
+
+class BaseFormatter(Generic[T]):
```
Comment (webknjaz) [inline lib/ansiblelint/formatters/__init__.py:14]: Not sure if it's necessary. It looks like it's needed to just use custom type @ `__main__`

File: lib/ansiblelint/formatters/__init__.py [hunk 4]
```diff
@@ -42,6 +44,9 @@ def _format_path(self, path: Union[str, Path]) -> str:
         # Use os.path.relpath 'cause Path.relative_to() misbehaves
         return os.path.relpath(path, start=self._base_dir)
 
+    def format(self, match: "MatchError", colored: bool = False) -> str:
+        return str(match)
+
```
Comment (webknjaz) [inline lib/ansiblelint/formatters/__init__.py:49]: I don't think this is related to reducing the complexity in main

Comment (ssbarnea) [reply lib/ansiblelint/formatters/__init__.py:49]: It forces inherited objects to respect type, this change is as much about typing as reducing complexity. I will make it clear in description, as original commit message did not endup being copied and the subject line is too short.

File: lib/ansiblelint/__main__.py (hunk 2 above)
Comment (ssbarnea) [reply lib/ansiblelint/__main__.py:74]: I tried different approaches and they did not work. If you know a better way to do it, one that is passing all our testing, lets do it in a follow-up.

File: lib/ansiblelint/formatters/__init__.py (hunk 3 above)
Comment (ssbarnea) [reply lib/ansiblelint/formatters/__init__.py:14]: See my previous comment.

File: lib/ansiblelint/__main__.py [hunk 5]
```diff
@@ -19,13 +19,12 @@
 # OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 # THE SOFTWARE.
 """Command line implementation."""
-
```
Comment (webknjaz) [inline lib/ansiblelint/__main__.py:22]: Unrelated. Also it's better to have a visual separator.

File: lib/ansiblelint/__main__.py (hunk 1 above)
Comment (webknjaz) [reply lib/ansiblelint/__main__.py:65]: This request still stands.

File: lib/ansiblelint/__main__.py (hunk 2 above)
Comment (webknjaz) [reply lib/ansiblelint/__main__.py:74]: I can take a look in a week.

**Pull request closed without merging**